src/Libraries/Microsoft.Extensions.AI.Abstractions/Embeddings/BinaryEmbedding.cs

```diff
@@ -0,0 +1,121 @@
+    /// <param name="vector">The embedding vector this embedding represents.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="vector"/> is <see langword="null"/>.</exception>
+    public BinaryEmbedding(BitArray vector)
+    {
+        _vector = Throw.IfNull(vector);
+    }
+
+    /// <summary>Gets or sets the embedding vector this embedding represents.</summary>
+    [JsonConverter(typeof(VectorConverter))]
+    public BitArray Vector
+    {
+        get => _vector;
+        set => _vector = Throw.IfNull(value);
+    }
+
+    /// <inheritdoc />
+    [JsonIgnore]
+    public override int Dimensions => _vector.Length;
+
+    /// <summary>Provides a <see cref="JsonConverter{BitArray}"/> for serializing <see cref="BitArray"/> instances.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class VectorConverter : JsonConverter<BitArray>
+    {
+        /// <inheritdoc/>
+        public override BitArray Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            _ = Throw.IfNull(typeToConvert);
+            _ = Throw.IfNull(options);
+
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new JsonException("Expected StartObject token");
+            }
+
+            int? length = null;
+            byte[]? bits = null;
+
+            while (reader.Read())
+            {
+                if (reader.TokenType == JsonTokenType.EndObject)
+                {
+                    break;
+                }
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    throw new JsonException("Expected PropertyName token");
+                }
+
+                string? propertyName = reader.GetString();
+                _ = reader.Read();
+
+                switch (propertyName)
+                {
+                    case "length":
+                        length = reader.GetInt32();
+                        break;
+
+                    case "bits":
+                        bits = reader.GetBytesFromBase64();
+                        break;
+
+                    default:
+                        reader.Skip();
+                        break;
+                }
+            }
+
+            if (length is null || bits is null)
+            {
+                throw new JsonException("Missing length or bits property");
+            }
+
+            return new BitArray(bits) { Length = length.Value };
+        }
+
+        /// <inheritdoc/>
+        public override void Write(Utf8JsonWriter writer, BitArray value, JsonSerializerOptions options)
+        {
+            _ = Throw.IfNull(writer);
+            _ = Throw.IfNull(value);
+            _ = Throw.IfNull(options);
+
+            writer.WriteStartObject();
+
+            writer.WritePropertyName("length");
```

Comment:
Is this implementing a schema that is standard for binary embeddings or is it just for convenience in roundtripping within .NET applications?

Reply:
Given that we're using base64 I think transmitting the value modulo 8 would suffice.

Reply:
> Is this implementing a schema that is standard for binary embeddings or is it just for convenience in roundtripping within .NET applications?

The latter.

> Given that we're using base64 I think transmitting the value modulo 8 would suffice.

Can you elaborate?

Reply:
> Can you elaborate?

Not super important. I was simply pointing out that transmitting the value of `length % 8` would suffice since the base64 representation already incorporates the number of bytes that fit the embedding.